ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala

```diff
@@ -16,174 +15,226 @@ import com.daml.lf.transaction.GlobalKey
+            languageVersion: LanguageVersion,
+            allowedLanguageVersions: VersionRange[LanguageVersion],
+        ): String =
+          LfError.Package
+            .AllowedLanguageVersion(packageId, languageVersion, allowedLanguageVersions)
+            .message
+
+        case class Error(
+            packageId: Ref.PackageId,
+            languageVersion: language.LanguageVersion,
+            allowedLanguageVersions: VersionRange[language.LanguageVersion],
+        )(implicit
+            val loggingContext: ContextualizedErrorLogger
+        ) extends LoggingTransactionErrorImpl(
+              cause = buildCause(packageId, languageVersion, allowedLanguageVersions)
+            )
+      }
+
+      @Explanation(
+        """This error occurs if a package referred to by a command fails validation. This should not happen as packages are validated when being uploaded."""
+      )
+      @Resolution("Contact support.")
+      object PackageValidationFailed
+          extends ErrorCode(
+            id = "PACKAGE_VALIDATION_FAILED",
+            ErrorCategory.MaliciousOrFaultyBehaviour,
+          ) {
+        case class Reject(validationErrorCause: String)(implicit
+            loggingContext: ContextualizedErrorLogger
+        ) extends LoggingTransactionErrorImpl(
+              cause = validationErrorCause
+            )
+      }
     }
 
-    @Explanation(
-      "This rejection is given when a request might not been processed and a time-out was reached."
-    )
-    @Resolution(
-      "Retry for transient problems. If non-transient contact the operator as the time-out limit might be to short."
-    )
-    object RequestTimeOut
-        extends ErrorCode(
-          id = "REQUEST_TIME_OUT",
-          ErrorCategory.DeadlineExceededRequestStateUnknown,
-        ) {
-      case class Reject(message: String, override val definiteAnswer: Boolean)(implicit
-          loggingContext: ContextualizedErrorLogger
-      ) extends LoggingTransactionErrorImpl(
-            cause = message
-          )
+    object Preprocessing extends ErrorGroup {
+      @Explanation("""This error occurs if a command fails during interpreter pre-processing.""")
+      @Resolution("Inspect error details and correct your application.")
+      object PreprocessingFailed
+          extends ErrorCode(
+            id = "COMMAND_PREPROCESSING_FAILED",
+            ErrorCategory.InvalidIndependentOfSystemState,
+          ) {
+        case class Reject(
+            err: LfError.Preprocessing.Error
+        )(implicit
+            loggingContext: ContextualizedErrorLogger
+        ) extends LoggingTransactionErrorImpl(
+              cause = err.message
+            )
+      }
     }
-  }
 
-  object ReadErrors extends ErrorGroup() {
+    object Interpreter extends ErrorGroup {
+      @Explanation("""This error occurs if a Daml transaction fails during interpretation.""")
+      @Resolution("This error type occurs if there is an application error.")
```

Comment:
This does not tell the user how to resolve the issue. An alternative could be `Ensure that your application generates a valid Daml transaction.`.